kwok/cloudprovider/cloudprovider.go

```diff
@@ -133,49 +132,39 @@ func (c CloudProvider) toNode(nodeClaim *v1beta1.NodeClaim) (*v1.Node, error) {
 	//nolint
 	newName = fmt.Sprintf("%s-%d", newName, rand.Uint32())
 
-	capacityType := v1beta1.CapacityTypeOnDemand
-	requirements := scheduling.NewNodeSelectorRequirementsWithMinValues(nodeClaim.
-		Spec.Requirements...)
-	if requirements.Get(v1beta1.CapacityTypeLabelKey).Has(v1beta1.CapacityTypeSpot) {
-		capacityType = v1beta1.CapacityTypeSpot
-	}
+	requirements := scheduling.NewNodeSelectorRequirementsWithMinValues(nodeClaim.Spec.Requirements...)
```

Comment:
The capacity type is specified inside the instance type offering.  It's not guaranteed that people are going to even create a set of offerings for their cluster that have spot instances in them, so we can't just assume that the capacity type will be spot.

(For example, in the kwok/examples/instance_types.json file, the only offerings that are specified are on-demand)

Reply:
@njtran In (old) L139, the check is looking to see if the _nodeClaim_ allows/requests spot, not if there are spot instances available from the _offering_.  If the embedded list of instance types only has on-demand instances in it, the check in (old) L163 will fail because there is no offering that supports spot, however we've unconditionally said we will return a spot instance.

This is relevant even if the generated list of instance types includes spot, because not every zone has spot instances in it.  For instance, in us-east-1, r7a.xlarge is only available as spot in 5 of the 6 AZs, which means if we unconditionally declare up front that we need a spot instance based on the node claim, and then later select the AZ that doesn't have spot instances, it will fail even though the zone potentially could have launched an on-demand instance.

Does that make sense?

Reply:
So the NodeClaim represents a list of potential options to the cloud provider for a given list of instance types, already considering any out of capacity errors. The NodeClaim itself shouldn't have the spot requirement here if there are no spot offerings available. This is just the function to translate the nodeclaim into a node, which is a selection of already vetted choices. Let me know if this makes sense.

Reply:
@njtran I don't believe this is true in the kwok provider, at least.  I created a branch off of `main` where the only change I made was commenting out the `v1beta1.CapacityTypeSpot` option on L90 of helpers.go (where we construct the list of static instance types), and then demonstrate that submitting a vanilla nginx pod causes karpenter to fail to scale up, because of the check that we're discussing.  Specifically, it fails with `failed to find offering spot/test-zone-a/c-128x-amd64-linux`.

I filed #1144 as a bug and include steps to reproduce the behaviour.

Reply:
Ah, thanks for pushing back on me on this. I got my wires crossed. Let's tag this issue that you opened as part of the PR description and we can close this out. Thanks!